storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,780 @@
+            if ((currentAssignment == null) ^ (container == null)) {
+                throw new IllegalArgumentException("Container and current assignment must both be null, or neither can be null");
+            }
+            
+            if (currentAssignment == null) {
+                state = MachineState.EMPTY;
+            } else {
+                state = MachineState.RUNNING;
+            }
+            
+            this.startTime = System.currentTimeMillis();
+            this.newAssignment = newAssignment;
+            this.pendingLocalization = null;
+            this.pendingDownload = null;
+            this.profileActions = new HashSet<>();
+            this.pendingStopProfileActions = new HashSet<>();
+        }
+        
+        public DynamicState(final MachineState state, final LocalAssignment newAssignment,
+                final Container container, final LocalAssignment currentAssignment,
+                final LocalAssignment pendingLocalization, final long startTime,
+                final Future<Void> pendingDownload, final Set<TopoProfileAction> profileActions, 
+                final Set<TopoProfileAction> pendingStopProfileActions) {
+            this.state = state;
+            this.newAssignment = newAssignment;
+            this.currentAssignment = currentAssignment;
+            this.container = container;
+            this.pendingLocalization = pendingLocalization;
+            this.startTime = startTime;
+            this.pendingDownload = pendingDownload;
+            this.profileActions = profileActions;
+            this.pendingStopProfileActions = pendingStopProfileActions;
+        }
+        
+        public String toString() {
+            StringBuffer sb = new StringBuffer();
+            sb.append(state);
+            sb.append(" msInState: ");
+            sb.append(Time.currentTimeMillis() - startTime);
+            if (container != null) {
+                sb.append(" ");
+                sb.append(container);
+            }
+            return sb.toString();
+        }
+
+        public DynamicState withNewAssignment(LocalAssignment newAssignment) {
+            return new DynamicState(this.state, newAssignment,
+                    this.container, this.currentAssignment,
+                    this.pendingLocalization, this.startTime,
+                    this.pendingDownload, this.profileActions,
+                    this.pendingStopProfileActions);
+        }
+        
+        public DynamicState withPendingLocalization(LocalAssignment pendingLocalization, Future<Void> pendingDownload) {
+            return new DynamicState(this.state, this.newAssignment,
+                    this.container, this.currentAssignment,
+                    pendingLocalization, this.startTime,
+                    pendingDownload, this.profileActions,
+                    this.pendingStopProfileActions);
+        }
+        
+        public DynamicState withPendingLocalization(Future<Void> pendingDownload) {
```

Comment:
nit, could return: withPendingLocalization(this.pendingLocalization, pendingDownload);